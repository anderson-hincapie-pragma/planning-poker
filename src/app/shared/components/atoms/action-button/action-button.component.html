<button class="a-action-button" [disabled]="buttonDisabled" data-testid="action-button" (click)="clickEvent($event)">{{text}}</button>