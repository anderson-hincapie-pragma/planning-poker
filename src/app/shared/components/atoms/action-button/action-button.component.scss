:host {
    display: contents;
}
.a-action-button {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    padding: 16px 24px;
    width: 159px;
    height: 40px;
    background: #F6F7FC;
    border: none;
    border-radius: 50px;
    font-family: 'Poppins';
    font-style: normal;
    font-weight: 600;
    font-size: 16px;
    line-height: 1;
    color: #330072;

    &:hover {
        background: #BB65FF;
        color: #ffffff;
    }

    &:disabled {
        background: rgba(157, 157, 161, 0.3);
        pointer-events: none;
        color: rgba(157, 157, 161, 0.6);
    }
}