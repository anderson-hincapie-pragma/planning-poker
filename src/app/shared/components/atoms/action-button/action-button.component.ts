import { Component, EventEmitter, Input, OnDestroy, OnInit, Output } from '@angular/core';
import { Subject, Subscription, throttleTime } from 'rxjs';

@Component({
  selector: 'app-action-button',
  templateUrl: './action-button.component.html',
  styleUrls: ['./action-button.component.scss']
})
export class ActionButtonComponent implements OnInit, OnDestroy {

  @Input()
  public text: string = "text";

  @Input()
  public buttonDisabled: boolean = false;

  @Input()
  public throttleTime = 500;

  @Output()
  public actionButtonClick: EventEmitter<MouseEvent> = new EventEmitter();
  private clicks = new Subject();
  private subscription?: Subscription;

  ngOnInit() {
    this.subscription = this.clicks.pipe(
      throttleTime(this.throttleTime)
    ).subscribe(e => this.emitThrottledClick(e as any));
  }

  ngOnDestroy(): void {
    this.subscription?.unsubscribe();
  }

  emitThrottledClick(e: MouseEvent) {
    this.actionButtonClick.emit(e);
  }

  clickEvent(event: MouseEvent) {
    this.clicks.next(event);
  }

}
