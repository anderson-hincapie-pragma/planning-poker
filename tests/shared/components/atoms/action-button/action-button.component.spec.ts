import { ComponentFixture, TestBed } from '@angular/core/testing';

import { ActionButtonComponent } from '@shared/components/atoms/action-button/action-button.component';
import { DebugElement } from '@angular/core';

describe('ActionButtonComponent', () => {
  let component: ActionButtonComponent;
  let fixture: ComponentFixture<ActionButtonComponent>;
  let compiled: HTMLElement;
  let debugElement: DebugElement;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [ActionButtonComponent]
    });
    fixture = TestBed.createComponent(ActionButtonComponent);
    debugElement = fixture.debugElement;
    compiled = fixture.nativeElement;
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('shoud match snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  it('should show default title', () => {
    let button = compiled.querySelector('[data-testid="action-button"]');
    expect(button?.textContent).toEqual(component.text);
  });

  it('should fire clickEvent on click', async () => {
    await fixture.whenStable();
    jest.spyOn(component, 'emitThrottledClick');
    jest.spyOn(component, 'clickEvent');
    let button = compiled.querySelector('[data-testid="action-button"]') as HTMLButtonElement;
    button.click();
    expect(component.clickEvent).toHaveBeenCalledTimes(1);
    expect(component.emitThrottledClick).toHaveBeenCalledTimes(1);
  });

  it('should throttle multiple clicks depending of throttleTime', async () => {
    await fixture.whenStable();
    jest.spyOn(component, 'emitThrottledClick');
    jest.spyOn(component, 'clickEvent');
    let button = compiled.querySelector('[data-testid="action-button"]') as HTMLButtonElement;
    button.click();
    expect(component.clickEvent).toHaveBeenCalledTimes(1);
    expect(component.emitThrottledClick).toHaveBeenCalledTimes(1);
    await new Promise<void>((resolve, reject) => setTimeout(() => resolve(), (component.throttleTime / 3) * 1));
    button.click();
    expect(component.clickEvent).toHaveBeenCalledTimes(2);
    expect(component.emitThrottledClick).toHaveBeenCalledTimes(1);
    await new Promise<void>((resolve, reject) => setTimeout(() => resolve(), (component.throttleTime / 3) * 1));
    button.click();
    expect(component.clickEvent).toHaveBeenCalledTimes(3);
    expect(component.emitThrottledClick).toHaveBeenCalledTimes(1);
    await new Promise<void>((resolve, reject) => setTimeout(() => resolve(), (component.throttleTime / 3) * 1));
    button.click();
    expect(component.clickEvent).toHaveBeenCalledTimes(4);
    expect(component.emitThrottledClick).toHaveBeenCalledTimes(2);
  });

});
