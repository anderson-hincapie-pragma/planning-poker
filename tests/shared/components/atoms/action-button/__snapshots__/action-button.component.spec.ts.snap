// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButtonComponent shoud match snapshot 1`] = `
<div
  id="root1"
>
  <button
    class="a-action-button"
    data-testid="action-button"
  >
    text
  </button>
</div>
`;
